src/main/java/com/google/devtools/build/lib/rules/proto/ProtoCompileActionBuilder.java

```diff
@@ -557,9 +562,13 @@ static CustomCommandLine createCommandLineFromToolchains(
   @VisibleForTesting
   static void addIncludeMapArguments(
       CustomCommandLine.Builder commandLine,
+      RuleContext ruleContext,
       @Nullable NestedSet<Artifact> protosInDirectDependencies,
       NestedSet<Artifact> transitiveImports) {
-    commandLine.addAll(VectorArg.of(transitiveImports).mapped(EXPAND_TRANSITIVE_IMPORT_ARG));
+    for (Artifact proto : transitiveImports) {
+      commandLine.addFormatted(
+          "-I%s=%s", getPathIgnoringRepository(ruleContext, proto), proto.getExecPathString());
```

Comment:
Why do you compute the path a particular proto file is accessible at here instead of the rule in whose `srcs` attribute the proto file is? That way, you would not need a `RuleContext`. 

And fortunately, `ProtoSourcesProvider` (or your new provider) already providers a place to do that. What I would do is to make `ProtoSourcesProvider#getTransitiveProtoSources()` a `NestedSet` of not artifacts, but of import path - artifact pairs and compute the import path of a `.proto` file in the `proto_library()` rule that owns it. Alternatively, if you think that the version without import path munging is the common case, have two methods: one for default import paths (without `import_prefix` or `strip_import_prefix` attributes) and one for changed import paths.

Thinking again, the latter is preferable because it makes merging this code a bit easier due to the existing users of that method.

Reply:
I was looking for where `protoc` was invoked, then trying to modify code closest to that invocation (for testing). It's certainly likely that this logic belongs in some other part of the build graph.

Are there any constraints on what types a `NestedSet` can contain? I'm not very familiar with Java, but IIRC there's no generic tuple types -- do I need to find a `ProtoSourcePath` type to handle that? Does it need to implement some sort of hashable interface to be put in a NestedSet?